### DISCUSSION TITLE
Raise npm min version to 8.x.x

### DISCUSSION THREAD

**Discussion by skjnldsv** [nextcloud/standards#6] [General]
To not pollute the main ticket.
from @vinicius73 

All current installation of Node 16 will use NPM v8
And NPM v8 have some important particularities who look relevant.
> npm@8 will refuse to install in node versions it doesn't work with
https://github.com/npm/rfcs/issues/445

This particular behaviour is very important to prevent more correlated issues about NPM and Node version in any project.

The first LTS version of Node 16 uses NPM8

Version | LTS | Date | npm
-- | -- | -- | --
Node.js 16.15.1 | Gallium | 2022-06-01 | 8.11.0
Node.js 16.15.0 | Gallium | 2022-04-26 | 8.5.5
Node.js 16.14.2 | Gallium | 2022-03-17 | 8.5.0
Node.js 16.14.1 | Gallium | 2022-03-16 | 8.5.0
Node.js 16.14.0 | Gallium | 2022-02-08 | 8.3.1
Node.js 16.13.2 | Gallium | 2022-01-10 | 8.1.2
Node.js 16.13.1 | Gallium | 2021-12-01 | 8.1.2
Node.js 16.13.0 | Gallium | 2021-10-26 | 8.1.0
Node.js 16.12.0 |   | 2021-10-20 | 8.1.0

**Comment by skjnldsv** [comment]
> > npm@8 will refuse to install in node versions it doesn't work with
> 
> [npm/rfcs#445](https://github.com/npm/rfcs/issues/445)

Which is all good, npm8 supports 12 and above :shrug: 

> npm@8 drops support for node<12
> npm@8 will refuse to install in node versions it doesn't work with

You can still install older node16 versions. I wanted to stay safe.

**Comment by skjnldsv** [comment]
From @st3iny 

Also, some distros still ship NPM 7 by default (e.g. Debian).

**Reply by st3iny** [reply]
To elaborate on my point: Ideally, we should all use NVM to manage our node versions independently of our distros. However, we should also pay attention to default NPM versions of popular distros to make it easier for new contributors.

**Comment by vinicius73** [comment]
I've checked the Debian Node and NPM versions.
As expected, Debian Stable/LTS version do not use Node v16, also the testing versions uses Node v16 and NPM v8.

| Debian | Released At | Status | Node | NPM |
| ------ | ----------- | ------ | ---- | -- |
| [Stretch v9](https://wiki.debian.org/DebianStretch) | 2017-06 | Old Stable | [Node v4.8](https://packages.debian.org/source/stretch/nodejs) | - |
| [Buster v10](https://wiki.debian.org/DebianBuster) | 2019-07 | Old Stable | [Node v10.24](https://packages.debian.org/source/buster/nodejs) | [NPM v5.8](https://packages.debian.org/source/buster/npm) |
| [Bullseye v11](https://wiki.debian.org/DebianBullseye) | 2021-08 | Stable | [Node v12.22](https://packages.debian.org/source/bullseye/nodejs) | [NPM v7.5](https://packages.debian.org/source/bullseye/npm) |
| [Bookworm](https://wiki.debian.org/DebianBookworm) | - | Testing | [Node v16.15](https://packages.debian.org/source/bookworm/nodejs) | [NPM v8.13](https://packages.debian.org/source/bookworm/npm) |
| [Sid](https://wiki.debian.org/DebianUnstable) | - | Experimental | [Node v16.15](https://packages.debian.org/source/sid/nodejs) | [NPM v8.13](https://packages.debian.org/source/bookworm/npm) |

In [Node download pages](https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions) we are redirected to [NodeSource](https://github.com/nodesource/distributions/blob/master/README.md) binaries and install scripts.

After install Node v16, we got NPM 8

```sh
curl -fsSL https://deb.nodesource.com/setup_lts.x | bash -
apt-get install -y nodejs

node --version 
#> v16.15.1

npm --version 
#> 8.11.0
```

Do not look reasonable to me, keep the NPM v7, when it is not used in any combination of those examples.

----

## Docker images test

> For my suprise, even if `Ubuntu 22.04` uses Node v14, its ships `NPM v8` inspead of `NPM v7`
> Fedora 36 uses Node 16 and NPM v8

```sh
apt update
apt install nodejs -y
apt install npm -y
``` 

### Stretch v9

> docker run -it --rm debian:stretch bash

```sh
nodejs --version 
#> v4.8.2

npm --version 
#> npm: command not found
```

### Buster v10

> docker run -it --rm debian:buster bash

```sh
nodejs --version 
#> v10.24.0

npm --version 
#> 5.8.0
```

### Bullseye v11

> docker run -it --rm debian:bullseye bash

```sh
nodejs --version 
#> v12.22.12

npm --version 
#> 7.5.2
```


### Debian Bookworm

> docker run -it --rm debian:bookworm bash

```sh
node --version 
#> v16.15.1

npm --version 
#> 8.13.1
```

### Debian Sid

> docker run -it --rm debian:sid bash

```sh
node --version 
#> v16.15.1

npm --version 
#> 8.13.2
```


### Ubuntu 22.04

> docker run -it --rm ubuntu:22.04 bash

```sh
node --version 
#> v12.22.9

npm --version 
#> 8.5.1
```

### Fedora 36

> docker run -it --rm fedora:36 bash

```sh
dfn update
dnf install node
```

```sh
node --version 
#> v16.14.0

npm --version 
#> 8.3.1
```

### OpenSuse Leap 15

> docker run -it --rm opensuse/leap:15 bash

```sh
zypper install nodejs16
zypper install npm
```

```sh
node --version 
#> v16.14.1

npm --version 
#> 8.5.0
```

> OpenSuse allows to choose any version of Node.
> Onse I've installed npm16 package where available, and install npm v8

**Reply by vinicius73** [reply]
I'm used to using [`n`](https://www.npmjs.com/package/n) to manage my Node versions.
Every time when I change Node version, N uses the correlated version of NPM

```sh
sudo n lts
#> installed : v16.15.1 (with npm 8.11.0)
```

```sh
sudo n 14
#> installed : v14.19.3 (with npm 6.14.17)
```

**Reply by skjnldsv** [reply]
What if I want a specific version of node 16? :shrug: 
```sh
10:42 [admin@workstation] ~% nvm install 16.2.0                                                                                                                                                                                                                                                                     -3- 
Downloading and installing node v16.2.0...
Downloading https://nodejs.org/dist/v16.2.0/node-v16.2.0-linux-x64.tar.xz...
######################################################################################################################################################################################################################################################### 100,0%
Computing checksum with sha256sum
Checksums matched!
Now using node v16.2.0 (npm v7.13.0)

```

**Reply by st3iny** [reply]
Wow, Bullseye is still on Node 12?! Thanks for your in-depth research.

**Reply by max-nextcloud** [reply]
> What if I want a specific version of node 16? shrug

wouldn't `npm install npm@8` do the trick on node v16.2.0?

**Reply by skjnldsv** [reply]
> > What if I want a specific version of node 16? shrug
> 
> wouldn't `npm install npm@8` do the trick on node v16.2.0?

Yes, but that's beyond my point of "staying compatible with all node versions" :)
I'm also fine raising to npm 8, just stating that the versions out there are much wider and various that we thought :see_no_evil:

**Comment by jotoeri** [comment]
Ok - so i think i would also still keep npm **MIN**-version at 7. Just because there are Versions of Node 16, that install by default with npm 7 and that should work out of the box (and as far as i see, 'out of the box' is a common argument). Alternative would imo only be to increase the node version further to `^16.11`, thus to explicitely exclude all versions, that come with npm7.

( ... comment to be continued in #7 )

**Comment by vinicius73** [comment]
Can we move forward with NPM v8?
